File: modules/rclone/android/src/main/java/expo/modules/rclone/RcloneModule.kt
package expo.modules.rclone

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android registration for the rclone bridge.
 *
 * Executing a binary copied into writable app storage is blocked on modern
 * Android versions. The production implementation must link rclone's gomobile
 * AAR and call librclone in-process. Until that artifact is available, reject
 * explicitly instead of shipping a silent no-op or a broken process launcher.
 */
class RcloneModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Rclone")

    AsyncFunction("initialize") {
      throw IllegalStateException(
        "NotImplemented: RcloneModule.initialize requires an Android librclone AAR"
      )
    }

    AsyncFunction("rpcCall") { _: String, _: String ->
      throw IllegalStateException(
        "NotImplemented: RcloneModule.rpcCall requires an Android librclone AAR"
      )
    }

    AsyncFunction("startOAuthFlow") { _: String ->
      throw IllegalStateException(
        "NotImplemented: RcloneModule.startOAuthFlow requires an Android librclone AAR"
      )
    }

    AsyncFunction("exchangeOAuthCode") { _: String, _: String ->
      throw IllegalStateException(
        "NotImplemented: RcloneModule.exchangeOAuthCode requires an Android librclone AAR"
      )
    }
  }
}

File: modules/rclone/ios/RcloneModule.swift
import ExpoModulesCore

/**
 * iOS registration for the rclone bridge.
 *
 * The previous implementation attempted to launch a macOS command-line binary
 * with Foundation.Process. Process is unavailable on iOS and therefore made
 * every iOS native build fail. The production implementation must link
 * rclone's gomobile XCFramework and call librclone in-process. Until that
 * artifact is available, reject explicitly instead of shipping a silent no-op.
 */
public final class RcloneModule: Module {
  public func definition() -> ModuleDefinition {
    Name("Rclone")

    AsyncFunction("initialize") { (promise: Promise) in
      promise.reject(
        "RCLONE_NOT_IMPLEMENTED",
        "NotImplemented: RcloneModule.initialize requires an iOS librclone XCFramework"
      )
    }

    AsyncFunction("rpcCall") { (_: String, _: String, promise: Promise) in
      promise.reject(
        "RCLONE_NOT_IMPLEMENTED",
        "NotImplemented: RcloneModule.rpcCall requires an iOS librclone XCFramework"
      )
    }

    AsyncFunction("startOAuthFlow") { (_: String, promise: Promise) in
      promise.reject(
        "RCLONE_NOT_IMPLEMENTED",
        "NotImplemented: RcloneModule.startOAuthFlow requires an iOS librclone XCFramework"
      )
    }

    AsyncFunction("exchangeOAuthCode") { (_: String, _: String, promise: Promise) in
      promise.reject(
        "RCLONE_NOT_IMPLEMENTED",
        "NotImplemented: RcloneModule.exchangeOAuthCode requires an iOS librclone XCFramework"
      )
    }
  }
}
